db/db_compaction_test.cc

```diff
@@ -3490,6 +3455,63 @@ TEST_F(DBCompactionTest, WaitForCompactShutdownWhileWaiting) {
   ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
 }
 
+TEST_P(DBCompactionWaitForCompactTest, WaitForCompactWithOptionToFlush) {
+  // After creating enough L0 files that one more file will trigger the
+  // compaction, write some data in memtable. Calls WaitForCompact with option
+  // to flush. This will flush the memtable to a new L0 file which will trigger
+  // compaction. Lastly check for expected number of files, closing + reopening
+  // DB won't trigger any flush or compaction
+
+  int compaction_finished = 0;
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::BackgroundCompaction:AfterCompaction",
+      [&](void*) { compaction_finished++; });
+  // To make sure there's a flush/compaction debt
+  int unscheduled_flushes = 0;
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::MaybeScheduleFlushOrCompaction:BeforeSchedule",
+      [&](void* arg) { unscheduled_flushes += *static_cast<int*>(arg); });
+
+  int flush_finished = 0;
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::BackgroundCallFlush:End", [&](void*) { flush_finished++; });
+
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->LoadDependency(
+      {{"DBCompactionTest::WaitForCompactWithOptionToFlush",
+        "DBImpl::MaybeScheduleFlushOrCompaction:BeforeSchedule"}});
+
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+
+  // write to memtable (overlapping key with first L0 file), but no flush is
+  // needed at this point.
+  ASSERT_OK(Put(Key(0), "some random string"));
+  ASSERT_EQ(0, unscheduled_flushes);
+  ASSERT_EQ(0, compaction_finished);
+  ASSERT_EQ("2", FilesPerLevel());
+  TEST_SYNC_POINT("DBCompactionTest::WaitForCompactWithOptionToFlush");
+
+  // Wait for compaction to finish with option to flush
+  WaitForCompactOptions wait_for_impact_options;
+  wait_for_impact_options.flush = true;
```

Comment:
This test appears to be lacking a "control": a test that disabling the option disables the intended functionality. That can probably be fixed by leveraging the different test instantiations.